File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
public:
bool f(int sum,int ind,vector<int>& nums,vector<vector<int>>&dp){
    int n=nums.size();
    if(sum==0) return true;
    if(ind==n &&sum!=0) return false;
    if(dp[ind][sum]!=-1) return dp[ind][sum];
    bool not_pick=f(sum,ind+1,nums,dp);
    bool pick=false;
    if(nums[ind]<=sum) pick=f(sum-nums[ind],ind+1,nums,dp); //<= has to be the condition, misatake before was >=
    return dp[ind][sum]=pick||not_pick; //stores as 1 or 0 in dp
}
    bool canPartition(vector<int>& nums) {
        int n=nums.size();
        int total=0;
        for(auto num:nums){
            total+=num;
        }
        if(total%2!=0) return false; //odd total sum, no possible to split into 2 subsets.
        int target=total/2;
        vector<vector<int>>dp(n+1,vector<int>(target+1,-1));
        return f(target,0,nums,dp);
    }
};

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.py
class Solution:
    def summation(self,ind,target,nums,dp):
        if target==0:
            return True
        elif ind>=len(nums):
            return False
        if dp[ind][target]!=None:
            return dp[ind][target]
        take=False
        if nums[ind]<=target:
            take=self.summation(ind+1,target-nums[ind],nums,dp)
        not_take=self.summation(ind+1,target,nums,dp)
        dp[ind][target]= take or not_take
        return dp[ind][target]
    def canPartition(self, nums: List[int]) -> bool:
        sum=0
        for num in nums:
            sum+=num
        if sum%2!=0:
            return False
        else:
            sum=sum//2
        n=len(nums)
        #for sum=10 we need table for indices from 0 to 10 , rather than 0 to 9, hence (sum+1)
        dp=[[None]*(sum+1) for _ in range(n)]
        return self.summation(0,sum,nums,dp)

